File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        string temp;
        stack<string> st;
        
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                st.push(temp);
                temp = "";
            } else if (s[i] == ')') {
                reverse(temp.begin(), temp.end());
                if (!st.empty()) {
                    temp = st.top() + temp;
                    st.pop();
                }
            } else {
                temp += s[i];
            }
        }
        
        return temp;
    }
};

File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.java
class Solution {
        public String reverseParentheses(String s) {
        Stack<String> st = new Stack<>();
        String ans = "";
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                st.push(ans);
                ans = "";
            } else if (s.charAt(i) == ')') {
                StringBuilder reversed = new StringBuilder(ans).reverse();
                ans = st.pop() + reversed.toString();
            } else {
                ans += s.charAt(i);
            }
        }
        
        return ans;
    }
}
